pkg/store/proxy_heap.go

```diff
@@ -0,0 +1,780 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package store
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"io"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/go-kit/log/level"
+	grpc_opentracing "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/tracing"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+	"github.com/thanos-io/thanos/pkg/tracing"
+)
+
+// dedupResponseHeap is a wrapper around ProxyResponseHeap
+// that deduplicates identical chunks identified by the same labelset.
+// It uses a hashing function to do that.
+type dedupResponseHeap struct {
+	h *ProxyResponseHeap
+
+	responses []*storepb.SeriesResponse
+
+	previousResponse *storepb.SeriesResponse
+	previousNext     bool
+}
+
+func NewDedupResponseHeap(h *ProxyResponseHeap) *dedupResponseHeap {
+	return &dedupResponseHeap{
+		h:            h,
+		previousNext: h.Next(),
+	}
+}
+
+func (d *dedupResponseHeap) Err() error {
+	return d.h.Error()
+}
+
+func (d *dedupResponseHeap) At() *storepb.SeriesResponse {
+	defer func() {
+		d.responses = d.responses[:0]
+	}()
+	if len(d.responses) == 0 {
+		return nil
+	} else if len(d.responses) == 1 {
+		return d.responses[0]
+	}
+
+	chunkDedupMap := map[uint64]*storepb.AggrChunk{}
+
+	for _, resp := range d.responses {
+		if resp.GetSeries() == nil {
+			continue
+		}
+		for _, chk := range resp.GetSeries().Chunks {
+			for _, field := range []*storepb.Chunk{
+				chk.Raw, chk.Count, chk.Max, chk.Min, chk.Sum, chk.Counter,
+			} {
+				if field == nil {
+					continue
+				}
+				h := xxhash.Sum64(field.Data)
+
+				if _, ok := chunkDedupMap[h]; !ok {
+					chk := chk
+
+					chunkDedupMap[h] = &chk
+				}
+			}
+
+		}
+	}
+
+	// If no chunks were requested.
+	if len(chunkDedupMap) == 0 {
+		return d.responses[0]
+	}
+
+	finalChunks := make([]storepb.AggrChunk, 0, len(chunkDedupMap))
+	for _, chk := range chunkDedupMap {
+		finalChunks = append(finalChunks, *chk)
+	}
+
+	sort.Slice(finalChunks, func(i, j int) bool {
+		return finalChunks[i].Compare(finalChunks[j]) > 0
+	})
+
+	// Guaranteed to be a series because Next() only buffers one
+	// warning at a time that gets handled in the beginning.
+	lbls := d.responses[0].GetSeries().Labels
+
+	return storepb.NewSeriesResponse(&storepb.Series{
+		Labels: lbls,
+		Chunks: finalChunks,
+	})
+}
+
+func (d *dedupResponseHeap) Next() bool {
```

Comment:
Small nit, but can we have `Next` implementation before `At`? Usually we want to see Next impl first